Варіант №21

Посилання на Github: https://github.com/DSviatenko/machine-learning

In [17]:
import pandas as pd
import numpy as np

1. Прочитайте csv файл (метод read_csv) 

In [18]:
data = pd.read_csv("bestsellers with categories.csv", encoding='latin1')
data

Name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       Author  User Rating  Reviews  Price  Year        Genre  
0                    JJ Smith          4.7    17350      8  2016  Non Fiction  
1                Stephen King          4.6     2052     22  2011      Fiction  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3               George Orwell          4.7    21424      6  2017      Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
545               Jeff Kinney          4.9     9413      8  2019      Fiction  
546               Jen Sincero          4.7    14331      8  2016  Non Fiction  
547               Jen Sincero          4.7    14331      8  2017  Non Fiction  
548               Jen Sincero          4.7    14331      8  2018  Non Fiction  
549               Jen Sincero          4.7    14331      8  2019  Non Fiction  

[550 rows x 7 columns]

2. Виведіть перші 10 рядків (метод head)

In [19]:
print(data.head(10))

                                                Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   
5      A Dance with Dragons (A Song of Ice and Fire)   
6  A Game of Thrones / A Clash of Kings / A Storm...   
7                     A Gentleman in Moscow: A Novel   
8      A Higher Loyalty: Truth, Lies, and Leadership   
9                          A Man Called Ove: A Novel   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fic

3. Виведіть розміри датасету (метод shape)

In [20]:
print(data.shape)

(550, 7)


**Питання:** Про скільки книг зберігає дані датасет?

***Відповідь:*** *550*

4. Змініть назви всіх колонок: назва з маленької букви, без пробілів через підкреслення

In [21]:
data.columns = ['name', 'author', 'user_rating', 'reviews', 'price', 'year', 'genre'] 
data

name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       author  user_rating  reviews  price  year        genre  
0                    JJ Smith          4.7    17350      8  2016  Non Fiction  
1                Stephen King          4.6     2052     22  2011      Fiction  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3               George Orwell          4.7    21424      6  2017      Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
545               Jeff Kinney          4.9     9413      8  2019      Fiction  
546               Jen Sincero          4.7    14331      8  2016  Non Fiction  
547               Jen Sincero          4.7    14331      8  2017  Non Fiction  
548               Jen Sincero          4.7    14331      8  2018  Non Fiction  
549               Jen Sincero          4.7    14331      8  2019  Non Fiction  

[550 rows x 7 columns]

5. Вивести кількість пропусків у кожному стовпчику. 

In [22]:
data_miss = data.isna().sum()/len(data)*100
print(f"The amount of gaps in each column:\n", data_miss)

The amount of gaps in each column:
 name           0.0
author         0.0
user_rating    0.0
reviews        0.0
price          0.0
year           0.0
genre          0.0
dtype: float64


**Питання:** Чи є в якихось змінних пропуски? (Так / ні) 

***Відповідь:*** *Ні*

6. За допомогою методу unique  перевірте. Які є унікальні жанри в стовпці genre  

In [23]:
unique_genres = data['genre'].unique()
print("Unique genres:", unique_genres)

Unique genres: ['Non Fiction' 'Fiction']


**Питання:** Які є унікальні жанри? 

***Відповідь:*** *'Non Fiction' 'Fiction'*

7. Визначте максимальну, мінімальну, медіанну ціну за допомогою методів  max, min, mean, median

In [24]:
max_price = data['price'].max()
min_price = data['price'].min()
mean_price = data['price'].mean()
median_price = data['price'].median()

print(f"Max price is: ", max_price)
print(f"Min price is: ", min_price)
print(f"Mean price is: ", mean_price)
print(f"Median price is: ", median_price)

Max price is:  105
Min price is:  0
Mean price is:  13.1
Median price is:  11.0


**Питання:** Максимальна ціна?

***Відповідь:*** *105*

**Питання:** Мінімальна ціна? 

***Відповідь:*** *0*

**Питання:** Середня ціна? 

***Відповідь:*** *13.1*

**Питання:** Медіанна ціна? 

***Відповідь:*** *11*

8. Виконайте наступні завдання з пошуку і сортування даних

а) **Питання:** Який рейтинг у датасеті найвищий? 

In [25]:
max_rating = data['user_rating'].max()
print(f"Max price is: ", max_rating)

Max price is:  4.9


б) **Питання:** Скільки книг мають такий рейтинг?

In [26]:
amount_of_books = data[data['user_rating'] == max_rating]['name'].count()
print(f"The amount of books with this rating: ", amount_of_books)

The amount of books with this rating:  52


в) **Питання:** Яка книга має найбільше відгуків?

In [27]:
max_reviews = data['reviews'].idxmax()
name_of_book = data.loc[max_reviews, 'name']
print(f"Name of the book with most reviews: ", name_of_book)

Name of the book with most reviews:  Where the Crawdads Sing


г) **Питання:** З тих книг, що потрапили до Топ-50 у 2010 році, яка книга найдорожча ?

In [28]:
top_50_in_2010 = data[(data['year'] == 2010) & (data['user_rating'] <= 50)]
most_expensive_book_2010 = top_50_in_2010.loc[top_50_in_2010['price'].idxmax()]
name_of_most_expensive_in_2010 = most_expensive_book_2010['name']
price_of_most_expensive_in_2010 = most_expensive_book_2010['price']
print("The most expensive book:")
print(f"Name: ", name_of_most_expensive_in_2010)
print(f"Price: ", price_of_most_expensive_in_2010)

The most expensive book:
Name:  Publication Manual of the American Psychological Association, 6th Edition
Price:  46


ґ) **Питання:** Скільки книг жанру Fiction потрапили до Топ-50 у 2012 році

In [29]:
top_50_in_2012 = data[(data['year'] == 2012) & (data['user_rating'] <= 50)]
fiction_count = len(top_50_in_2012[top_50_in_2012['genre'] == 'Fiction'])
print(f'The number of books with the genre "Fiction" among the top 50 in 2012:', fiction_count)

The number of books with the genre "Fiction" among the top 50 in 2012: 21


д) **Питання:** Скільки книг з рейтингом 4.9 потрапило до рейтингу у 2010 та 2011 роках (використовуйте | або метод isin)?

In [30]:
years = (data['year'] == 2010) | (data['year'] == 2011)
rating_4_9 = data[years & (data['user_rating'] == 4.9)]
amount_of_books = len(rating_4_9)
print(f"Amount of books with 4.9 rating in 2010 and 2011: ", amount_of_books)

Amount of books with 4.9 rating in 2010 and 2011:  1


9. Вивести максимальну і  мінімальну ціну для жанру Fiction I NonFiction (за допомогою методів  groupby та agg, для підрахунку мінімальних та максимальних значень використовуйте max та min). Використовуйте тількі потрібні вам стовпці. 

In [31]:
genre_group = data.groupby('genre')
result = genre_group['price'].agg(['min', 'max'])
print(result)

             min  max
genre                
Fiction        0   82
Non Fiction    0  105


***Висновок:***  *на практичній роботі я поглибив знання з мови програмування Python, використовуючи бібліотеку "pandas". Протягом роботи я виконував різні завдання: зчитавання CSV файлів та обробки таблиць, виконував пошук мінімальних та максимальних елементів таблиц, виконував підрахунки даних відповідно до заданих обмежень(фільтрів)*